server/text_generation_server/models/ct2_causal_lm.py

```diff
@@ -0,0 +1,381 @@
+        out_dir = (
+            Path(HUGGINGFACE_HUB_CACHE)
+            / f"ct2models-{model_id.replace('/','--')}--{ct2_compute_type}"
+        )
+
+        if not os.path.exists(out_dir / "model.bin"):
+            try:
+                converter = ctranslate2.converters.TransformersConverter(
+                    model_id,
+                    activation_scales=None,
+                    load_as_float16=True,
+                    revision=None,
+                    low_cpu_mem_usage=True,
+                    trust_remote_code=trust_remote_code,
+                )
+                converter.convert(
+                    output_dir=out_dir,
+                    vmap=None,
+                    quantization=ct2_compute_type,
+                    force=True,
+                )
+            except Exception as ex:
+                raise ValueError(
+                    f"conversion with ctranslate2 for {model_id} failed : Error {ex}"
+                )
+        if not os.path.exists(out_dir / "model.bin"):
+            raise ValueError(
+                f"no ctranslate2 for {model_id} found after conversion in {out_dir}"
+            )
+
+        # Start CT2
+        self.ct2_model = ctranslate2.Generator(
+            str(out_dir), device=self.ct2_device, compute_type=ct2_compute_type
+        )
+
+        class DummyModel(torch.nn.Module):
+            def __init__(self, *args, **kwargs) -> None:
+                super().__init__(*args, **kwargs)
+                self.config = AutoConfig.from_pretrained(
+                    model_id, revision=revision, trust_remote_code=trust_remote_code
+                )
+
+        model = DummyModel()
+
+        if tokenizer.pad_token_id is None:
+            if model.config.pad_token_id is not None:
+                tokenizer.pad_token_id = model.config.pad_token_id
+            elif model.config.eos_token_id is not None:
+                tokenizer.pad_token_id = model.config.eos_token_id
+            elif tokenizer.eos_token_id is not None:
+                tokenizer.pad_token_id = tokenizer.eos_token_id
+            else:
+                tokenizer.add_special_tokens({"pad_token": "[PAD]"})
+
+        super(CT2CausalLM, self).__init__(
+            model=model,
+            tokenizer=tokenizer,
+            requires_padding=True,
+            dtype=torch.int32,
+            device=torch.device("cuda"),
+        )
+
+    @property
+    def batch_type(self) -> Type[CausalLMBatch]:
+        return CausalLMBatch
+
+    def decode(self, generated_ids: List[int]) -> str:
+        return self.tokenizer.decode(
+            generated_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False
+        )
+
+    # def forward_slowtokenize_ct2(
+    #     self, all_input_ids,
+    # ) -> Tuple[torch.Tensor, List[Tuple[torch.Tensor, torch.Tensor]]]:
+    #     # Model Forward, by copy between cpu and cuda
+    #     tokens_in  = [self.tokenizer.convert_ids_to_tokens(i) for i in all_input_ids]
```

Comment:
This code also works quite fast, however moving the tokens from cuda -> cpu -> cuda. On the other hand does not require padding.